### PR TITLE
Unify training scripts to learn from demonstrations

### PR DESCRIPTION
**Pull request by AdamGleave**
WIP. So far have just refactored train_adversarial to make structure clearer. Will start seeing what's common with train_bc next, and then eventually add MCE IRL and density modeling.

### CODE REVIEW

File: src/imitation/algorithms/mce_irl.py [hunk 1]
```diff
@@ -135,6 +136,50 @@ def squeeze_r(r_output: th.Tensor) -> th.Tensor:
...
+        assert pi.ndim == 3, "expected three-dimensional policy"
+        assert np.allclose(pi.sum(axis=2), 1), "policy not normalized"
+        assert np.all(pi >= 0), "policy has negative probabilities"
+        self.pi = pi
+        self.rng = rng or np.random
+
+    def predict(
+        self,
+        observation: np.ndarray,
+        state: Optional[np.ndarray] = None,
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/algorithms/mce_irl.py:162]: Thanks for flagging this discrepancy. I think I might just be confused about what tabular MDPs observation is meant to be, then. In a POMDP, the policy would see the observation; the reward is typically a function of state, but for a learned reward it has to be a function of observation. Sounds like in the current implementation this is swapped: policy sees state but reward sees observation?

File: src/imitation/scripts/common/train.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    policy_cls = base.FeedForward32Policy
+    policy_kwargs = {}
+
+    # Evaluation
+    n_episodes_eval = 50  # Num of episodes for final mean ground truth return
+
+    # TODO(adam): should we separate config into a separate file?
+    # or just disable F841 on this whole file?
+    _ = locals()  # quieten flake8
+    del _
```
[the remark this answers is not in the document]
Comment (qxcv) [reply src/imitation/scripts/common/train.py:53]: I mean this:

```python
# This is redundant:
_ = locals()
del _

# Instead:
locals()
```

(disclaimer: I only tested this quickly in my editor and am not certain it will work for all versions of flake8)

File: src/imitation/algorithms/mce_irl.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (qxcv) [reply src/imitation/algorithms/mce_irl.py:162]: Yes. This is standard for MCE IRL: the "observations" only exist so you can get the benefits of function approximation for your reward function (namely, it forces semantically similar states to have similar reward). Given that you are doing planning to obtain your policy anyway, there is no need for function approximation in the policy.

File: src/imitation/scripts/common/train.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/scripts/common/train.py:53]: Thanks, this works, will make that change!

File: src/imitation/algorithms/base.py [hunk 3]
```diff
@@ -166,6 +167,16 @@ def set_demonstrations(self, demonstrations: AnyTransitions) -> None:
                 `TransitionKind` instance, or a Sequence of Trajectory objects.
         """
 
+    # TODO(adam): this will cause minor headache for MCE IRL and density baseline
+    # which do not train a policy by default. Could return None for these?
+    # Or just have them train when calling this. Not so bad for MCE IRL
+    # (can compute it analytically in reasonable time frame), more awkward
+    # for density baseline.
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/algorithms/base.py:174]: I wanted `.policy()` so that we can have a consistent interface, with the same way to evaluate the resulting policy across all algorithms. However, the interface is different enough in other ways I'm not sure how advantageous this ends up being. Implementing `policy()` turned out OK for MCE and density, so I'm inclined to leave this as-is, but I think it'd be reasonable to switch to `NotImplementedError` in the future if we have methods where supporting this is awkward.

File: src/imitation/algorithms/mce_irl.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/algorithms/mce_irl.py:186]: Thanks for the explanation. We need to stick to the SB3 naming convention for this function for compatibility with any callees that are using keyword arguments. But I've added a discussion of this in the docstring to avoid confusion.

File: src/imitation/scripts/train_bc.py [hunk 4]
```diff
@@ -3,160 +3,116 @@
...
     )
+    # TODO(adam): have BC support timesteps instead of epochs/batches?
     model.train(
         n_epochs=n_epochs,
         n_batches=n_batches,
         log_interval=log_interval,
         log_rollouts_venv=venv,
         log_rollouts_n_episodes=log_rollouts_n_episodes,
     )
+    # TODO(adam): add checkpointing to BC?
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/scripts/train_bc.py]: This PR is already long so I plan to do this in a new branch

**Review by qxcv — CHANGES_REQUESTED** [review]
A few small changes, otherwise LGTM. I didn't check the configs closely because I assumed they were all copy-pasted. I noticed some of these were missing `# noqa: F841` and `locals()` tricks though—is there some other way that you got flake8 to stop complaining about them?

File: src/imitation/algorithms/dagger.py [hunk 5]
```diff
@@ -306,30 +305,34 @@ def __init__(
...
+            bc_trainer.action_space,
         )
+        self.bc_trainer = bc_trainer
+        self.bc_trainer.logger = custom_logger
+
+    def __getstate__(self):
+        """Return state excluding non-pickleable objects."""
+        d = dict(self.__dict__)
+        del d["venv"]
+        del d["_logger"]
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/algorithms/dagger.py:334]: Did this ever work? In general environments just aren't pickleable. I don't want to break existing functionality, but I'm loathe to add anything new in this (already massive) PR.

We already have a custom save method, adding a custom load method that requires a venv and logger would be easy enough to do. WDYT?
